### PR TITLE
feat(filter-stream): Add a generateChecksums option

### PR DESCRIPTION
**Pull request by zvin**
This option will update the blockmap adding a checksum to every Range.

Change-Type: minor

### CODE REVIEW

File: lib/filter-stream.js [hunk 1]
```diff
@@ -122,13 +127,17 @@ FilterStream.HIGH_WATER_MARK = 64 * 1024
...
-    if( range.checksum !== digest ) {
+    if( this.options.verify && (range.checksum !== digest) ) {
       error = new Error( `Invalid checksum for range [${range.startLBA},${range.endLBA}], bytes ${range.start}-${range.end}` )
       error.checksum = digest
       error.range = range
       this.emit( 'error', error )
     }
 
+    if ( this.options.calculateChecksums ) {
+      range.range.checksum = digest
```
Comment (lurch) [inline lib/filter-stream.js]: Hmm? How come `this.options.verify` checks for `(range.checksum !== digest)`, but `this.options.calculateChecksums` sets `range.range.checksum = digest` ?  (apologies if that's a stupid question)

File: lib/filter-stream.js [hunk 2]
```diff
@@ -11,6 +11,7 @@ var BlockMap = require( './blockmap' )
  * @param {BlockMap} blockMap - the block map
  * @param {Object} [options] - options
  * @param {Boolean} [options.verify=true] - verify range checksums
+ * @param {Boolean} [options.calculateChecksums=false] - calculate range checksums and update blockmap
```
Comment (lurch) [inline lib/filter-stream.js]: How do `options.verify` and `options.calculateChecksums` interact? Is it expected that `verify` will always be used with blockmaps that **do** have checksums, and `calculateChecksums` will always be used with blockmaps that **don't** have checksums?
What happens if `options.verify` and `options.calculateChecksums` are both set to true at the same time? (should there be a unit-test for that?)

Comment (zvin) [reply lib/filter-stream.js]: This wouldn't make sense.
`verify` and `calculateChecksums` are exclusive.

File: lib/filter-stream.js (hunk 1 above)
Comment (zvin) [reply lib/filter-stream.js]: `ReadRange`s are created from `Range`s and their checksum is copied.
See https://github.com/resin-io-modules/blockmap/blob/master/lib/read-range.js#L20 .

File: lib/filter-stream.js (hunk 2 above)
Comment (jhermsmeier) [reply lib/filter-stream.js]: Should we rename `calculateChecksums` to `updateChecksums` or `generateChecksums`?

Comment (zvin) [reply lib/filter-stream.js]: changed to `generateChecksums`

Comment (zvin) [reply lib/filter-stream.js]: @lurch the constructor will throw an error if both `verify` and `generateChecksums` are set.

**Review by jhermsmeier — APPROVED** [review]

**Pull request merged**